File: homestay_app/lib/models/payment.dart
class Payment {
  final int id;
  final int bookingId;
  final double amount;
  final String method; // VNPay, Momo, BankTransfer
  final String status; // Pending, Completed, Failed, Refunded
  final String? transactionId;
  final DateTime? completedAt;
  final DateTime createdAt;

  Payment({
    required this.id,
    required this.bookingId,
    required this.amount,
    required this.method,
    required this.status,
    this.transactionId,
    this.completedAt,
    required this.createdAt,
  });

  factory Payment.fromJson(Map<String, dynamic> json) {
    return Payment(
      id: json['id'] ?? 0,
      bookingId: json['bookingId'] ?? 0,
      amount: (json['amount'] ?? 0).toDouble(),
      method: json['method'] ?? '',
      status: json['status'] ?? 'Pending',
      transactionId: json['transactionId'],
      completedAt: json['completedAt'] != null 
          ? DateTime.parse(json['completedAt']) 
          : null,
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'bookingId': bookingId,
      'amount': amount,
      'method': method,
      'status': status,
      'transactionId': transactionId,
      'completedAt': completedAt?.toIso8601String(),
      'createdAt': createdAt.toIso8601String(),
    };
  }

  String get statusDisplay {
    switch (status) {
      case 'Pending':
        return 'Chờ thanh toán';
      case 'Completed':
        return 'Đã thanh toán';
      case 'Failed':
        return 'Thất bại';
      case 'Refunded':
        return 'Đã hoàn tiền';
      default:
        return status;
    }
  }

  String get methodDisplay {
    switch (method) {
      case 'VNPay':
        return 'VNPay';
      case 'Momo':
        return 'MoMo';
      case 'BankTransfer':
        return 'Chuyển khoản';
      default:
        return method;
    }
  }
}

class AppNotification {
  final int id;
  final String userId;
  final String type;
  final String title;
  final String message;
  final String? relatedUrl;
  final bool isRead;
  final DateTime createdAt;

  AppNotification({
    required this.id,
    required this.userId,
    required this.type,
    required this.title,
    required this.message,
    this.relatedUrl,
    this.isRead = false,
    required this.createdAt,
  });

  factory AppNotification.fromJson(Map<String, dynamic> json) {
    return AppNotification(
      id: json['id'] ?? 0,
      userId: json['userId'] ?? '',
      type: json['type'] ?? '',
      title: json['title'] ?? '',
      message: json['message'] ?? '',
      relatedUrl: json['relatedUrl'],
      isRead: json['isRead'] ?? false,
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'type': type,
      'title': title,
      'message': message,
      'relatedUrl': relatedUrl,
      'isRead': isRead,
      'createdAt': createdAt.toIso8601String(),
    };
  }
}

File: CSDL/Models/Payment.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using User = Nhom1.Models.User;

namespace Nhom1.Models
{
    public class Payment
    {
        public int Id { get; set; }

        [Required(ErrorMessage = "Transaction ID is required")]
        [StringLength(100, ErrorMessage = "Transaction ID cannot exceed 100 characters")]
        public string TransactionId { get; set; } = string.Empty;

        [Column(TypeName = "decimal(18,2)")]
        [Range(0.01, double.MaxValue, ErrorMessage = "Amount must be greater than 0")]
        public decimal Amount { get; set; }

        public PaymentMethod PaymentMethod { get; set; }
        public PaymentStatus Status { get; set; } = PaymentStatus.Pending;

        [StringLength(500, ErrorMessage = "Notes cannot exceed 500 characters")]
        public string? Notes { get; set; }

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        public DateTime? UpdatedAt { get; set; }
        public DateTime? CompletedAt { get; set; }

        // Foreign keys
        [Required]
        public string UserId { get; set; } = string.Empty;

        [Required]
        public int BookingId { get; set; }

        // Navigation properties
        [ForeignKey("UserId")]
        public virtual User User { get; set; } = null!;
        
        [ForeignKey("BookingId")]
        public virtual Booking Booking { get; set; } = null!;

        // Computed properties
        [NotMapped]
        public string PaymentMethodDisplay => PaymentMethod switch
        {
            PaymentMethod.MoMo => "MoMo",
            PaymentMethod.VNPay => "VNPay",
            PaymentMethod.PayPal => "PayPal", 
            PaymentMethod.Stripe => "Stripe",
            PaymentMethod.BankTransfer => "Bank Transfer",
            PaymentMethod.Free => "Free",
            _ => "Unknown"
        };

        [NotMapped]
        public string StatusDisplay => Status switch
        {
            PaymentStatus.Pending => "Pending",
            PaymentStatus.Processing => "Processing",
            PaymentStatus.Completed => "Completed",
            PaymentStatus.Failed => "Failed",
            PaymentStatus.Cancelled => "Cancelled",
            PaymentStatus.Refunded => "Refunded",
            _ => "Unknown"
        };

        [NotMapped]
        public bool IsCompleted => Status == PaymentStatus.Completed;

        [NotMapped]
        public bool CanBeRefunded => Status == PaymentStatus.Completed && PaymentMethod != PaymentMethod.Free;

        [NotMapped]
        public bool IsFailed => Status == PaymentStatus.Failed;

        [NotMapped]
        public TimeSpan? ProcessingTime => CompletedAt.HasValue ? CompletedAt.Value - CreatedAt : null;
    }

    public enum PaymentMethod
    {
        MoMo = 0,
        VNPay = 1,
        PayPal = 2,
        Stripe = 3,
        BankTransfer = 4,
        Free = 5 // Phương thức thanh toán miễn phí
    }

    public enum PaymentStatus
    {
        Pending = 0,
        Processing = 1,
        Completed = 2,
        Failed = 3,
        Cancelled = 4,
        Refunded = 5
    }
}
